### PR TITLE
store: use db iterators for pruning and range-based queries

### PR DESCRIPTION
**Pull request by cmwaters**
## Description

With order preserving keys, db's can now take advantage of iterators for pruning and range-based queries like finding the `base()` and `height()` of the block store

This modifies:
- Block Store
- Light Client Store

modifications to the state store has some tricky logic and has been separated to a different PR (#5864)

### CODE REVIEW

File: light/store/db/db.go [hunk 1]
```diff
@@ -238,37 +243,30 @@ func (s *dbs) Prune(size uint16) error {
...
 
-	s.size -= uint16(pruned)
+	if wErr := b.Set(s.sizeKey(), marshalSize(size)); wErr != nil {
+		err = fmt.Errorf("failed to persist size: %w", wErr)
+	}
 
-	if wErr := s.db.SetSync(s.sizeKey(), marshalSize(size)); wErr != nil {
-		return fmt.Errorf("failed to persist size: %w", wErr)
+	// 4) write batch deletion to disk
+	err = b.WriteSync()
```
Comment (cmwaters) [inline light/store/db/db.go:267]: Do we want to apply the same logic to pruning the light client store as to the block and state store where after x amount of light blocks we flush to disk

Comment (erikgrinaker) [reply light/store/db/db.go:267]: Depends on how much data there is to delete in a typical worst case. Batches are typically buffered in memory, so may OOM the node if they grow too large.

Comment (cmwaters) [reply light/store/db/db.go:267]: A typical light block for a 100 validator network would be 15-17kB

Comment (erikgrinaker) [reply light/store/db/db.go:267]: The batches don't store the blocks themselves, only the keys scheduled for deletion. So it would be average key size multiplied by worst-case number of keys. Some databases may also have limits on maximum batch sizes.

Comment (cmwaters) [reply light/store/db/db.go:267]: I can't see this being a problem then

File: store/store.go [hunk 2]
```diff
@@ -33,60 +31,92 @@ The store can be assumed to contain all contiguous blocks between base and heigh
...
-	return &BlockStore{
-		base:   bs.Base,
-		height: bs.Height,
-		db:     db,
-	}
+	return &BlockStore{db}
 }
 
 // Base returns the first known contiguous block height, or 0 for empty block stores.
 func (bs *BlockStore) Base() int64 {
```
Comment (cmwaters) [inline store/store.go:43]: I've noticed that the current design pattern with this store is to suppress any errors from arising by either panicking or failing silently. We don't even have any logging with the block store (as opposed to say the evidence pool). Are we fine with continuing with this? In contrast the state store bubbles up errors for almost every method.

Comment (erikgrinaker) [reply store/store.go:43]: We should return errors. But that will probably have ripple-effects to many other parts of the code, so it's better to change that in a separate PR to avoid too much noise (and since it's a logically separate change).

File: store/store.go [hunk 3]
```diff
@@ -245,82 +275,181 @@ func (bs *BlockStore) LoadSeenCommit(height int64) *types.Commit {
...
+	if err != nil {
+		return pruned, fmt.Errorf("pruning error at key %X: %w", iter.Key(), err)
+	}
+
+	return pruned, nil
+}
+
+// batchDelete is a generic function for deleting a range of values based on the lowest
+// height up to but excluding retainHeight
+func (bs *BlockStore) batchDelete(start []byte, end []byte) error {
```
Comment (cmwaters) [inline store/store.go:397]: This is quite a generic function (i.e. state store does the exact same thing). It would probably make more sense to extract this. My first thought is to add it to the tm-db? or to the lib directory. Or should I just repeat myself (I remember someone telling me there's nothing wrong with that). What are people's thoughts on this?

Comment (cmwaters) [reply store/store.go:397]: Hmm actually it might be not so generic as I thought. State store has a few nuances. I will just repeat myself for now

Comment (erikgrinaker) [reply store/store.go:397]: Might make sense to extract this out to a helper function, but maybe design it in such a way that it can handle those "special" cases too, e.g. by using callback functions.

Comment (cmwaters) [reply store/store.go:397]: I guess the main reason why I would have two distinct functions is because the block store iterates forward and the state store iterates backwards (because state store has this entire dependency dynamic)

File: light/store/db/db.go [hunk 4]
```diff
@@ -30,7 +29,7 @@ type dbs struct {
 // want to use one DB with many light clients).
 func New(db dbm.DB, prefix string) store.Store {
 
-	lightStore := &dbs{db: db, prefix: prefix}
+	lightStore := &dbs{db: dbm.NewPrefixDB(db, []byte(prefix))}
```
Comment (erikgrinaker) [inline light/store/db/db.go:32]: If the caller wants to use a prefix, they should just pass a `PrefixDB` as `db`. It implements the `DB` interface, so it's entirely transparent and the light store won't need any special handling of it at all.

File: light/store/db/db.go [hunk 5]
```diff
@@ -214,7 +219,7 @@ func (s *dbs) LightBlockBefore(height int64) (*types.LightBlock, error) {
 // left.
 //
 // Safe for concurrent use by multiple goroutines.
-func (s *dbs) Prune(size uint16) error {
+func (s *dbs) Prune(size uint16) (err error) {
```
Comment (erikgrinaker) [inline light/store/db/db.go:222]: I'm not really a fan of named return values, it's too easy to get them wrong (e.g. by shadowing the variable and returning `nil` instead of what you think you're returning).

File: light/store/db/db.go [hunk 6]
```diff
@@ -238,37 +243,30 @@ func (s *dbs) Prune(size uint16) error {
...
-	// 3) Update size.
+	// 3) // update size
 	s.mtx.Lock()
-	defer s.mtx.Unlock()
+	s.size = size
+	s.mtx.Unlock()
 
-	s.size -= uint16(pruned)
+	if wErr := b.Set(s.sizeKey(), marshalSize(size)); wErr != nil {
+		err = fmt.Errorf("failed to persist size: %w", wErr)
```
Comment (erikgrinaker) [inline light/store/db/db.go:263]: Should we not return this error? We're replacing it on the next line below.

File: store/store.go [hunk 7]
```diff
@@ -33,60 +31,92 @@ The store can be assumed to contain all contiguous blocks between base and heigh
...
 func (bs *BlockStore) Base() int64 {
-	bs.mtx.RLock()
-	defer bs.mtx.RUnlock()
-	return bs.base
+	iter, err := bs.db.Iterator(
+		blockMetaKey(1),
+		blockMetaKey(1<<63-1),
+	)
+	if err != nil {
+		return 0
```
Comment (erikgrinaker) [inline store/store.go:49]: We shouldn't ignore errors. If you don't want to deal with error propagation in this PR then at least panic, but this should really return an error before the next release.

File: store/store.go [hunk 8]
```diff
@@ -33,60 +31,92 @@ The store can be assumed to contain all contiguous blocks between base and heigh
...
+	}
+	defer iter.Close()
+
+	if iter.Valid() {
```
Comment (erikgrinaker) [inline store/store.go:53]: This needs to check for iterator errors too (after this block, since an errored iterator is invalid).

File: store/store.go [hunk 9]
```diff
@@ -245,82 +275,181 @@ func (bs *BlockStore) LoadSeenCommit(height int64) *types.Commit {
...
+	if err != nil {
+		return pruned, err
 	}
-	base := bs.base
-	bs.mtx.RUnlock()
-	if height < base {
-		return 0, fmt.Errorf("cannot prune to height %v, it is lower than base height %v",
-			height, base)
+
+	if err := bs.batchDelete(blockPartKey(1, 0), blockPartKey(height, 0)); err != nil {
```
Comment (erikgrinaker) [inline store/store.go:295]: Might as well start from height 0 in all of these ranges.

File: store/store.go [hunk 10]
```diff
@@ -245,82 +275,181 @@ func (bs *BlockStore) LoadSeenCommit(height int64) *types.Commit {
...
+		err = proto.Unmarshal(bz, pbbm)
+		if err != nil {
+			panic(fmt.Errorf("unmarshal to tmproto.BlockMeta: %w", err))
 		}
-		if err := batch.Delete(blockCommitKey(h)); err != nil {
-			return 0, err
+
+		blockMeta, err := types.BlockMetaFromProto(pbbm)
+		if err != nil {
+			panic(fmt.Errorf("error from proto blockMeta: %w", err))
```
Comment (erikgrinaker) [inline store/store.go:343]: Why are we panicing when the function returns an error?

File: store/store.go [hunk 11]
```diff
@@ -382,16 +511,6 @@ func (bs *BlockStore) SaveBlock(block *types.Block, blockParts *types.PartSet, s
...
-	// Done!
-	bs.mtx.Lock()
-	bs.height = height
-	if bs.base == 0 {
-		bs.base = height
-	}
-	bs.mtx.Unlock()
-
-	// Save new BlockStoreState descriptor. This also flushes the database.
-	bs.saveState()
```
Comment (erikgrinaker) [inline store/store.go:394]: Note how this flushes the database. We're not using `SetSync()` above, for performance reasons, which means the OS is just buffering the writes in memory. We need to make sure it's flushed to disk at the end of this method, either by changing the last `Set()` call to `SetSync()` or doing a final explicit sync write.

File: store/store.go [hunk 12]
```diff
@@ -245,82 +275,181 @@ func (bs *BlockStore) LoadSeenCommit(height int64) *types.Commit {
...
+	if err := bs.batchDelete(seenCommitKey(1), seenCommitKey(height)); err != nil {
+		return pruned, err
+	}
+
+	return pruned, nil
+}
+
+// pruneBlockMetaAndHashes prunes all block metas from the base height to the retainHeight
+// but before each, loads the hash from the block meta to also prune from the block hash table
+func (bs *BlockStore) pruneBlockMetaAndHashes(retainHeight int64) (uint64, error) {
```
Comment (erikgrinaker) [inline store/store.go:312]: It's a bit sad that we have to duplicate all of this. If `batchDelete` took a callback function `func(key, value []byte, batch dbm.Batch) error` then we could pass `nil` in the typical case, and a function that extracts the block meta and adds the hash delete to the batch for the block case.

**Review by erikgrinaker — COMMENTED** [review]
Good stuff, love seeing `BlockStoreState` go! This needs a changelog entry as well.

Also, there seems to be some related test failures.

File: store/store.go (hunk 7 above)
Comment (cmwaters) [reply store/store.go:49]: Yeah I am going to add error handling in a separate PR. This just keeps it with the current behavior for now

File: store/store.go (hunk 10 above)
Comment (cmwaters) [reply store/store.go:343]: good point. I was just copying behavior from another part of the code

File: store/store.go (hunk 8 above)
Comment (cmwaters) [reply store/store.go:53]: Just to clear up my understanding, how should certain errors thrown by an iterator be handled:
```
it, err := bs.db.Iterator()  -> we should panic
err := it.Error() -> we should return an error if we can else ignore
err := it.Close() -> we should return an error if we can else ignore
```

File: store/store.go (hunk 12 above)
Comment (cmwaters) [reply store/store.go:312]: Okay I can try add something like this

File: store/store.go (hunk 11 above)
Comment (cmwaters) [reply store/store.go:394]: Sure. Just to clarify, databases will also flush stuff to disk when they think the time is right, not just when we specify it's time to flush by calling `SetSync()`?

File: light/store/db/db.go (hunk 6 above)
Comment (cmwaters) [reply light/store/db/db.go:263]: yup good catch

File: store/store.go (hunk 11 above)
Comment (erikgrinaker) [reply store/store.go:394]: This is mostly managed by the operating system, not the database, but yeah it will flush to disk when appropriate. However, the OS may crash or lost power, in which case it won't make it to disk. This can cause problems, e.g. making a node double-sign if it crashed after submitting a previous vote but didn't persist a record of that vote to disk.

File: store/store.go (hunk 8 above)
Comment (erikgrinaker) [reply store/store.go:53]: In general, all errors should be returned. In certain cases you may get away with ignoring close errors (e.g. defers whee you're sure the consequences are negligible). Panics should generally only be used when returning an error is very inconvenient and the only failure modes are programmer error.

File: store/store.go (hunk 7 above)
Comment (erikgrinaker) [reply store/store.go:49]: Sure, but then panic instead.

File: store/store.go (hunk 8 above)
Comment (cmwaters) [reply store/store.go:53]: Okay, thanks. I will try address this all in a follow up PR

File: store/store.go [hunk 13]
```diff
@@ -285,116 +291,59 @@ func (bs *BlockStore) PruneBlocks(height int64) (uint64, error) {
...
-			iter, err = bs.db.Iterator(
-				blockMetaKey(1),
-				blockMetaKey(retainHeight),
-			)
-			if err != nil {
-				panic(err)
-			}
-			defer iter.Close()
+	// for the rest of the block data we just want to remove it
+	emptyClause := func(key, value []byte, batch dbm.Batch) error { return nil }
```
Comment (erikgrinaker) [inline store/store.go:324]: This isn't necessary, you can just pass `nil` to `batchDelete()´.

File: store/store.go [hunk 14]
```diff
@@ -285,116 +291,59 @@ func (bs *BlockStore) PruneBlocks(height int64) (uint64, error) {
...
 	return pruned, nil
 }
 
 // batchDelete is a generic function for deleting a range of values based on the lowest
 // height up to but excluding retainHeight
-func (bs *BlockStore) batchDelete(start []byte, end []byte) error {
+func (bs *BlockStore) batchDelete(
+	start []byte,
+	end []byte,
+	clause func(key, value []byte, batch dbm.Batch) error) (uint64, error) {
```
Comment (erikgrinaker) [inline store/store.go:346]: The term `clause` isn't immediately obvious to me -- what does it mean in this context? We should explain it in the comment as well.

Also, maybe move `(uint64, error)` to the next line, I found it a bit hard to parse this line since there's two different return value lists here.

**Review by erikgrinaker — APPROVED** [review]

File: store/store.go (hunk 14 above)
Comment (cmwaters) [reply store/store.go:346]: Yeah I also wasn't too sure if it explained what it was for. I was thinking that clause represented a secondary or auxiliary step in deletion. Perhaps something like `preprocess` might be more apt. What do you think?

Comment (erikgrinaker) [reply store/store.go:346]: Maybe just `process` -- preprocess sort of indicates that the results will be used as input to some other operation. Not really a great term either, but don't have any better ideas.

Comment (cmwaters) [reply store/store.go:346]: Basically this is a function that is called on a key/value pair before we do the deletion hence why I thought of the `pre` part. Perhaps `preDeletionHook`

Comment (erikgrinaker) [reply store/store.go:346]: Sure
